otherlibs/stdune/src/loc.ml

```diff
@@ -57,21 +57,27 @@ let pp_file_excerpt ~context_lines ~max_lines_to_print_in_full { start; stop } :
   let stop_c = stop.pos_cnum - start.pos_bol in
   let file = start.pos_fname in
   let pp_file_excerpt () =
-    let line_num = start.pos_lnum in
-    let line_num_str = string_of_int line_num in
-    let padding_width = String.length line_num_str in
     let open Result.O in
-    let* line =
-      Result.try_with (fun () -> Io.String_path.file_line file line_num)
+    let* res =
+      if start.pos_lnum <> stop.pos_lnum then Ok `Multiline
+      else
+        let line_num = start.pos_lnum in
+        let line_num_str = string_of_int line_num in
+        let padding_width = String.length line_num_str in
+        let* line =
+          Result.try_with (fun () -> Io.String_path.file_line file line_num)
+        in
+        let len = stop_c - start_c in
+        let open Pp.O in
+        Ok
+          (`Singleline
+            (pp_line padding_width (line_num_str, line)
+            ++ pp_left_pad (stop_c + padding_width + 3) (String.make len '^')
+            ++ Pp.newline))
```

Comment:
I would push this code under the `Ok Singleline` case below (similarly to the other case).